hal/src/rtl872x/rtl_sdk_support.cpp

```diff
@@ -190,10 +364,30 @@ void rtwRadioAcquire(RtwRadio r) {
         return;
     }
     if (radioStatus != RTW_RADIO_NONE) {
+        static std::once_flag once;
+        std::call_once(once, [](){
+            static rtl_wifi_fw_ram_alloc __attribute__((aligned(32))) info;
+            info.size = sizeof(info);
+            info.start = (uint32_t)&link_prebootloader_wifi_fw_ram_start;
+            info.end = (uint32_t)&link_prebootloader_wifi_fw_ram_end;
+            DCache_CleanInvalidate((uint32_t)&info, sizeof(info));
+            rtl_wifi_fw_resp resp = {};
+            int r = km0_km4_ipc_send_request(KM0_KM4_IPC_CHANNEL_GENERIC, KM0_KM4_IPC_MSG_WIFI_FW_INIT, &info, sizeof(info), [](km0_km4_ipc_msg_t* msg, void* context) -> void {
```

Comment:
What does the "wifi firmware" actually do? Why does it need to run on KM0?

Reply:
It manages low power wifi modes and mainly for our use-case prioritization between wifi and bt slots. Pretty much the main thing it does in simple terms is set up an interrupt at TBTT interval (100ms generally, depends on AP settings) and depending on the coexistence configuration may switch the RF frontend between WiFi and BLE peripheral. Of course it's a bit more complex than simply that.

Reply:
ok thanks for the general overview. I'm guessing wifi fw is running on KM0 for performance and hardware reasons? wifi fw is loaded into KM0 ROM?

Reply:
:shrug: That's how it's implemented by Realtek. Wifi fw is in `ambd_sdk/project/realtek_amebaD_va0_example/GCC-RELEASE/project_lp/asdk/lib/application/lib_wifi_fw.a`, it does use some of the ROM functions, but in general it just runs on KM0 through a set of threads / interrupt handlers. It can run XIP, but XIP is seriously broken as usual, so we run it out of KM4 SRAM (as we are out of KM0 SRAM).